src/note.rs

```diff
@@ -38,24 +54,29 @@ pub struct Note {
 
 impl Note {
     /// Derives the commitment to this note.
+    ///
+    /// Defined in [Zcash Protocol Spec § 3.2: Notes][notes].
+    ///
+    /// [notes]: https://zips.z.cash/protocol/nu5.pdf#notes
     pub fn commitment(&self) -> NoteCommitment {
-        todo!()
+        let g_d = self.recipient.g_d();
+
+        NoteCommitment::derive(
+            g_d.to_bytes(),
+            self.recipient.pk_d().to_bytes(),
+            self.value,
+            self.rho.0,
+            self.rseed.psi(),
+            (&self.rseed).into(),
+        )
     }
 
     /// Derives the nullifier for this note.
-    pub fn nullifier(&self, _: &FullViewingKey) -> Nullifier {
-        todo!()
+    pub fn nullifier(&self, fvk: &FullViewingKey) -> Nullifier {
+        Nullifier::derive(self.rho.0, self.rseed.psi(), self.commitment(), fvk.nk())
```

Comment:
I suggest making nk the first argument to `Nullifier::derive`, matching the spec.